### PR TITLE
event: fix Resubscribe deadlock when unsubscribing after inner sub ends

### PR DESCRIPTION
**Pull request by Inphi**
A goroutine oversees the lifetime of subscriptions handled by resubscriptions. This goroutine terminates when the subscription ends without any errors.  However, the resub goroutine needs to live long enough to read from the unsub channel. Otherwise, an Unsubscribe call deadlocks when writing to the unsub channel.

### CODE REVIEW

File: event/subscription_test.go [hunk 1]
```diff
@@ -154,3 +154,23 @@ func TestResubscribeWithErrorHandler(t *testing.T) {
...
+		return NewSubscription(func(unsubscribed <-chan struct{}) error {
+			select {
+			case <-time.After(2 * time.Second):
+				innerSubDone <- struct{}{}
+				return nil
+			case <-unsubscribed:
+				return nil
+			}
+		}), nil
+	})
```
Comment (holiman) [inline event/subscription_test.go:172]: I am not too familiar with how the semantics of subscriptions work in any depth, but I've been trying to figure out if this is 'correct' or not. The docs for NewSubscription says

>  NewSubscription runs a producer function as a subscription in a new goroutine. The
> channel given to the producer is closed when Unsubscribe is called. If fn returns an
> error, it is sent on the subscription's error channel.

In other words, `unsubscribed` will be closed when externalities wants the producer to stop. In this testcase, however, the producer just stops producing, and exiting without returning any error or closing any channel. And yeah, that intuitively seems like something that could lead to a deadlock. 

If the select is changed into
```golang
			select {
			case <-time.After(2 * time.Second):
				innerSubDone <- struct{}{}
				return errors.New("time to go")
			case <-unsubscribed:
				return nil
			}
```
Then the deadlock disappears.

So my thinking goes: this testcase is based on a flawed producer. But also, the `NewSubscription` documentation should mention something like "the producer must either exit with an error or keep listening to the given channel".

**Comment by fjl** [comment]
Maybe an easier fix would be adding a 1 buffer to `unsub`

File: event/subscription_test.go (hunk 1 above)
Comment (Inphi) [reply event/subscription_test.go:172]: I see that this [documentation](https://github.com/ethereum/go-ethereum/blob/d10a2f6ab727f79a0acff29c8147d54c5e4689ec/event/subscription.go#L47) already clarifies the intended behavior; producers are permitted to return nil errors. And `Subscription` explicitly checks for a nil error [here](https://github.com/ethereum/go-ethereum/blob/d10a2f6ab727f79a0acff29c8147d54c5e4689ec/event/subscription.go#L57C3-L57C3). While this behavior doesn't compose well with Resubscriptions, we should maintain the current API contract.

**Comment by Inphi** [comment]
@fjl Simplified the fix by making unsub buffered. Thanks for the suggestion.

File: event/subscription_test.go [hunk 2]
```diff
@@ -154,3 +154,23 @@ func TestResubscribeWithErrorHandler(t *testing.T) {
...
 }
+
+func TestResubscribeWithCompletedSubscription(t *testing.T) {
+	t.Parallel()
+
+	innerSubDone := make(chan struct{}, 1)
+	sub := ResubscribeErr(100*time.Millisecond, func(ctx context.Context, lastErr error) (Subscription, error) {
+		return NewSubscription(func(unsubscribed <-chan struct{}) error {
+			select {
+			case <-time.After(2 * time.Second):
```
Comment (fjl) [inline event/subscription_test.go:165]: It's weird to have a timeout here. You can achieve the necessary synchronization event with another channel, like so

```
quitInner := make(chan struct{})
...
    select {
    case <-quitInner:
...

close(quitInner)
<-innerSubDone
sub.Unsubscribe()
```

Comment (Inphi) [reply event/subscription_test.go]: good suggestion. Made the change.

**Pull request merged**